feat(@desktop/wallet): Added transaction address component

Cuteivist opened the pull request:
closes #10583

### What does the PR do

Changes introduce new components to
* Format address in specific way - `TransactionAddress`
* Base Tile component in Transaction Activity Details view - `TransactionActivityTile`
* Tile displaying  list of transaction addresses  - `TransactionAddressTile`

Currently only `TransactionAddress` component was added to storybook.

### Affected areas

Wallet - Transaction Activity Details

### Screenshot of functionality (including design for comparison)

Design: 
https://www.figma.com/file/FkFClTCYKf83RJWoifWgoX/Wallet-v2?node-id=12689-341126&t=zqIHTe3GG0EC5YOp-4

**NOTE**  - data on example image is mocked

![image](https://github.com/status-im/status-desktop/assets/11396062/88e5b863-be18-43b5-a2c7-1412ad7ed3d2)

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 4462039e | [#1](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-imports/job/PR-10665/1/) | 2023-05-12 13:13:22 | ~6 min | `tests/imports` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-imports/job/PR-10665/1//consoleText) |

Khushboo-dev-cpp commented on ui/imports/shared/controls/TransactionAddress.qml:
```diff
@@ -0,0 +1,134 @@
+                    if(!!root.address == false)
+                        return ""
+                    if (d.prefixAndAddress.prefix.length > 0) {
+                        return WalletUtils.colorizedChainPrefix(d.prefixAndAddress.prefix) + d.prefixAndAddress.address  + "asdasdaspdjsajdasdasjpasdjaspdas"
```
"asdasdaspdjsajdasdasjpasdjaspdas" leftover perhaps?

Cuteivist replied:
Sorry, you are right, fixed.

Khushboo-dev-cpp commented:
<img width="758" alt="image" src="https://github.com/status-im/status-desktop/assets/60327365/7df8c692-4d66-427e-ae22-77bab97b0ebd">
The gap between To and address seems more than design?

Khushboo-dev-cpp commented on ui/imports/shared/controls/TransactionAddressTile.qml:
```diff
@@ -0,0 +1,91 @@
+                width: parent.width
+                address: modelData
+                addressName: !!root.rootStore ? root.rootStore.getNameForAddress(address) : ""
+                // TODO remove me
```
leftover?

Cuteivist replied:
Sorry, you are right, fixed.

Khushboo-dev-cpp commented on ui/imports/shared/controls/TransactionAddress.qml:
```diff
@@ -0,0 +1,134 @@
+import QtQuick 2.15
```
I wonder why we couldn't reuse StatusListItem for this, maybe you have a thought behind this already

Khushboo-dev-cpp commented on ui/imports/shared/controls/TransactionAddressTile.qml:
```diff
@@ -0,0 +1,91 @@
+   The \c TransactionAddressTile can display list of addresses formatted in specific way.
+
+   \qml
+        TransactionAddressTile {
```
should be TransactionActivityTile ?

Khushboo-dev-cpp commented on ui/imports/shared/controls/TransactionAddressTile.qml:
```diff
@@ -0,0 +1,91 @@
+
+   The \c TransactionAddressTile can display list of addresses formatted in specific way.
+
+   \qml
```
the naming is a bit difficult to process as all of these only demonstrate Tx addres using it sses right?
Also I don't see the usage of this in the "current" Transaction Details View, so if you could do that, would be great.
Also having these in storybbook will help cause thats how we are heading ahead with implementation right now

Cuteivist replied:
I went with naming used there already - InformationTile.
My decision to create base component because we will be able to reuse it in whole screen.

Please look as an example here: https://www.figma.com/file/FkFClTCYKf83RJWoifWgoX/Wallet-v2?type=design&node-id=12689-341125&t=V0cyIC3m2qoRmcTU-0

Below addresses the "tiles" are have the same look, just different contents.

Only thing I think could make sense is to merge "TransactionAddress" with "TransactionAddressTile", but reason for having them independent is to be able to test them.

Cuteivist commented:
> <img alt="image" width="758" src="https://user-images.githubusercontent.com/60327365/237986166-7df8c692-4d66-427e-ae22-77bab97b0ebd.png"> The gap between To and address seems more than design?

Great catch! I fixed that too, updated main image.

caybro commented on ui/StatusQ/src/StatusQ/Components/StatusListItem.qml:
```diff
@@ -134,6 +134,7 @@ Rectangle {
     MouseArea {
         id: sensor
 
+        z: 1 // Gives ability to hide siblings under the MouseArea
```
I see yeah...

Cuteivist commented on src/app/modules/main/wallet_section/accounts/model.nim:
```diff
@@ -100,6 +100,6 @@ QtObject:
   proc getNameByAddress*(self: Model, address: string): string =
     for item in self.items:
-      if(item.address() == address):
+      if(toLower(item.address()) == toLower(address)):
```
I see there is https://nim-lang.org/docs/strutils.html#cmpIgnoreCase%2Cstring%2Cstring , but it will result in same calculations behind.

Cuteivist replied:
Looking deeper at the implementation looks that it will be faster, because it does by iterating.

Cuteivist replied:
Changed.

stefandunca commented on storybook/main.cpp:
```diff
@@ -58,7 +58,11 @@ int main(int argc, char *argv[])
                 "Storybook", 1, 0, "CacheCleaner", cleanerFactory);
 
+#ifdef Q_OS_WIN
+    const QUrl url(QUrl::fromLocalFile(QML_IMPORT_ROOT + QStringLiteral("/main.qml")));
```
Won't this also work for other platforms to simplify it and drop the compile time condition?

Cuteivist replied:
At this moment I'm not sure, because it would also affect MacOS devs.
This line doesn't change and I'm 100% sure that will cover all cases.
I think it's not worth investigating it.

stefandunca commented on ui/imports/shared/controls/TransactionActivityTile.qml:
```diff
@@ -0,0 +1,104 @@
+    /*!
+       \qmlproperty bool TransactionActivityTile::roundedCornersTop
+       This property shows rounded corners at the top of the background rectangle.
+       First tile should always have roundedCorners corners.
```
I'm wondering if there is an advantage/requirement for this approach over squared tiles and clip them within a clipped container with rounded corners.

Cuteivist replied:
Yes. It's because in design only first and second tile has rounded corners.
It might look that it could be solved by having border outside the tile, but issue is hovering. 
When I hover tile I want full fill, so first and last items will have filled color rectangle when hovering with radius on one side of the rectangle.
I think this solution is much better than for example using Shape, especially that it only loads for first and last tile.

![image](https://github.com/status-im/status-desktop/assets/11396062/9e7fb789-f462-41dd-b95e-881305f9f16b)

stefandunca replied:
For sure shape won't be an advantage. Still not sure why a simpler approach with the list embedded into the `Rectangle` component won't work; it doesn't worth considering it, though, when you have something in place.

e.g.

```qml
Rectangle {
  radius: 5
  clip: true
  ListView {
    delegate: TransactionActivityTile {}
    }
}

Cuteivist replied:
@stefandunca and @IvanBelyakoff 
Unfortunately clip doesn't work with radius in QML. I will remove that effect affecting rounded corners as it's API isn't perfect.
Whole solution might look like a hack, but it's best way to handle it pixel-perfect long term and it's proper way to handle such cases.

I will figure out frame component in another issue.

Cuteivist replied:
@IvanBelyakoff It's because of some optimizations on Qt side.
Generally I decided I will implement it during summary part where I will be using those tiles.

stefandunca reviewed (approved):
Good job. I left some questions inline

alexjba commented on ui/imports/shared/controls/TransactionAddress.qml:
```diff
@@ -0,0 +1,134 @@
+    QtObject {
+        id: d
+        readonly property var prefixAndAddress: Utils.splitToChainPrefixAndAddress(root.address)
+        readonly property bool isContactPubKeyValid: root.contactPubKey !== undefined && !!root.contactPubKey
```
`undefined` is a falsy value. This should do the trick
```suggestion
        readonly property bool isContactPubKeyValid: !!root.contactPubKey
```

Cuteivist replied:
Good point. Fixed

alexjba commented on ui/imports/shared/controls/TransactionAddress.qml:
```diff
@@ -0,0 +1,134 @@
+        id: d
+        readonly property var prefixAndAddress: Utils.splitToChainPrefixAndAddress(root.address)
+        readonly property bool isContactPubKeyValid: root.contactPubKey !== undefined && !!root.contactPubKey
+        readonly property var contactData: Utils.getContactDetailsAsJson(root.contactPubKey)
```
Would this binding work if the contact is removed? I guess removing the contact should update this UI in realtime. If this is the case we might need some `Connections`. There is a similar use case in `Chat/stores/RootStore.qml`.

An easy test would be to open this UI and remove the current user from the other account's contacts.

Cuteivist replied:
Good point! Contacts store was already passed to wallet details page, so I just added connections.

Thanks @alexjba

alexjba reviewed (commented):
Nice work! 👍 
Minor inline questions:

Cuteivist commented on storybook/pages/TransactionAddressTilePage.qml:
```diff
@@ -0,0 +1,125 @@
+            sourceComponent: Column {
+                width: parent.width
+                height: childrenRect.height
+                spacing: -1
```
Removed

Cuteivist commented on storybook/pages/TransactionAddressTilePage.qml:
```diff
@@ -0,0 +1,125 @@
+                width: parent.width
+                height: childrenRect.height
+                spacing: -1
+                TransactionAddressTile {
```
Changed, as mentioned in other comment

caybro reviewed (approved):
LGTM
